src/Mono.Android/Xamarin.Android.Net/AndroidMessageHandler.cs

```diff
@@ -1151,48 +1166,125 @@ void SetupSSL (HttpsURLConnection? httpsConnection, HttpRequestMessage requestMe
 				return;
 			}
 
-			var keyStore = InitializeKeyStore (out bool gotCerts);
-			keyStore = ConfigureKeyStore (keyStore);
-			var kmf = ConfigureKeyManagerFactory (keyStore);
+			var keyStore = GetConfiguredKeyStoreInstance (out bool gotTrustedCerts, out bool gotClientCerts);
+			var kmf = GetConfiguredKeyManagerFactory (keyStore, gotClientCerts);
 			var tmf = ConfigureTrustManagerFactory (keyStore);
 
-			if (tmf == null) {
-				// If there are no trusted certs, no custom trust manager factory or custom certificate validation callback
-				// there is no point in changing the behavior of the default SSL socket factory
-				if (!gotCerts && _serverCertificateCustomValidator is null)
-					return;
+			// If there are no trusted certs, no custom trust manager factory, no custom key manager factory, or custom
+			// certificate validation callback there is no point in changing the behavior of the default SSL socket factory
+			if (tmf == null && kmf == null && !gotTrustedCerts && _serverCertificateCustomValidator is null) {
+				return;
+			}
+
+			var context = SSLContext.GetInstance ("TLS");
+			var trustManagers = GetTrustManagers (tmf, keyStore, gotTrustedCerts, requestMessage);
+			context?.Init (kmf?.GetKeyManagers (), trustManagers, null);
+			httpsConnection.SSLSocketFactory = context?.SocketFactory;
+		}
 
+		KeyManagerFactory? GetConfiguredKeyManagerFactory (KeyStore? keyStore, bool gotClientCerts)
+		{
+			var kmf = ConfigureKeyManagerFactory (keyStore);
+			if (kmf == null && gotClientCerts) {
+				kmf = KeyManagerFactory.GetInstance ("PKIX");
+				kmf?.Init (keyStore, null);
+			}
+
+			return kmf;
+		}
+
+		KeyStore? GetConfiguredKeyStoreInstance (out bool gotTrustedCerts, out bool gotClientCerts)
```

Comment:
I'm not sure I like the implied semantics of the "wtf" path.

Assume that `ClientCertificateOptions=ClientCertificateOptions.Manual` and `ClientCertificates` is set to a non-empty collection.  This *presumably* means "the certificates should be used".

Execution eventually hits this method, which calls `KeyStore.GetInstance(KeyStore.DefaultType)`, which could return `null`.  *If* that happens, then:

 * `keyStore?.Load()` is ignored
 * `LoadClientCertificates(keyStore, out gotClientCerts)` becomes an effective no-op, *ignoring* the contents of `ClientCertificates` (okay, not quite, `X509Certificate2` instances will be created, other logic happens, but *nothing important* happens to the `keyStore` instance, as it's `null`)
 * …and then…???

*Should* this happen?  Of course not.

But if it *does* happen -- something goes FUBAR, reasons unknown -- then the downstream effect will be "as if" `ClientCertificates` weren't set at all, which in turn will likely result in a bunch of head scratching and cursing, or a security vulnerability.  (I have an overactive imagination.)

I would prefer that we instead require that things work:

```csharp
var keyStore = KeyStore.GetInstance (KeyStore.DefaultType) ??
    throw new InvalidOperationException($"KeyStore.GetInstance(\"{KeyStore.DefaultType}\") returned null!");
```

and then replace most of the `keyStore?.Whatever` with `keyStore.Whatever`, as we are *requiring* that `KeyStore` be non-`null`.

Reply:
Thans for pointing this out. The `GetInstance(DefaultType)` call never returns `null` AFAIK (at least on any existing Android version), so ensuring it is `null` will clean up the code and make the intent much clearer. We might also avoid weird silent errors at some later point if the behavior of the method changes.